apache/superset
# fix: Time shifts with different granularity for ECharts

### SUMMARY
There's an option in the Advanced Analytics section of Explore called Time Shift that allows an user to compare the query results with the same query dislocated in time by the amount specified in the control. The way this is implemented, is to fire a query for each specified Time Shift and later join the results with the original query. To make the join work the algorithm modifies the temporal column of the time shift results to match the one in the original query. The problem is that the algorithm is not taking into account that the time granularity and time shifts can influence on the temporal column and affect the resulting timestamps. For example, if the Time Grain is Quarter and Time Shift is 28 days, the results could have different timestamps but still belong to the same quarter. Another example is when the Time Grain is Week and after applying the time shifts, the results fall in a different year which may produce different week numbers for the timestamps. The objective of this PR is to fix these problems by generating an artificial join key that takes into consideration the Time Grain when joining the results.

The following table illustrates how the join keys are calculated depending on the Time Grain:

Time Grain | Join key
-- | --
Week | year-week
Month | year-month
Quarter | year-quarter
Year | year

This join key is only used when joining the results and is discarded in the end.

This PR also improves the Python codebase by creating a `TimeGrain` class and replacing all string literals with references to the new class.

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
<img width="1774" alt="Screenshot 2023-05-22 at 13 24 36" src="https://github.com/apache/superset/assets/70410625/e2bc854a-2f4b-4e5c-846b-fad2d2e1d711">
<img width="1777" alt="Screenshot 2023-05-22 at 13 26 02" src="https://github.com/apache/superset/assets/70410625/9b9d385b-5a9f-4bab-9481-6a7fa8327d1a">

### TESTING INSTRUCTIONS
1 - Choose a dataset with plenty of temporal data available
2 - Create a Line Chart (ECharts)
3 - Play with different Time Grains and Time Shifts
4 - Make sure the offsets are correctly calculated

### ADDITIONAL INFORMATION
